dashboard/app/models/user.rb

```diff
@@ -800,6 +800,38 @@ def User.track_script_progress(user_id, script_id)
     end
   end
 
+  # If appropriate, increases the level counts for the concept-difficulties
+  # associated with the completed level.
+  def User.track_proficiency(user_id, script_id, level_id)
+    level_concept_difficulty = LevelConceptDifficulty.where(level_id: level_id).first
+    if !level_concept_difficulty
+      return
+    end
+
+    # Exit early if a hint was used, as no proficiency is gained in this case.
+    hint_used = HintViewRequest.
+      where(user_id: user_id, script_id: script_id, level_id: level_id).
+      any?
+    if hint_used
+      return
+    end
+
+    retryable on: [Mysql2::Error, ActiveRecord::RecordNotUnique], matching: /Duplicate entry/ do
+      user_proficiency = UserProficiency.where(user_id: user_id).first_or_create!
+      user_proficiency.last_progress_at = Time.now
+
+      ConceptDifficulties::CONCEPTS.each do |concept|
+        if level_concept_difficulty.send(concept)
+          # TODO(asher): If possible, simplify this to use direct assignment via
+          # incrementing rather than going through setting attributes.
+          user_proficiency.attributes = {"#{concept}_d#{level_concept_difficulty.send(concept)}_count" => user_proficiency.send("#{concept}_d#{level_concept_difficulty.send(concept)}_count") + 1}
```

Comment:
this update bit requires enough knowledge about the details of user_proficiency that it seems like it should really go in a method on the model.